lib/ddtrace/context.rb

```diff
@@ -199,15 +250,15 @@ def check_finished_spans
       @finished_spans > 0 && @trace.length == @finished_spans
     end
 
-    def attach_sampling_priority
-      @current_root_span.set_metric(
+    def attach_sampling_priority(span)
```

Comment:
Because `Context#get` calls `reset` when returning a complete trace, we can't call this method after the fact.

I think a good solution is to make the `Context` "less smart", as methods like `#get` have a lot of logic in it that could be extracted to be ContextFlush responsibilities, like you suggested. That has to be done in a thread-safe way, respecting the possibility that Context is reused on the same thread after `reset` is called.

This to me looks like a separate, larger refactor in its own merit. I don't think the changes in PR go against that goal, on the contrary, but do I agree that this PR doesn't fully address the abstraction issue that currently existing with the `Context`.

I'd like to go ahead with the scope of this PR and schedule this much needed refactor for the near future.